staging/src/k8s.io/apiserver/pkg/storage/etcd3/lease_manager.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+)
+
+// leaseManager is used to manage leases requested from etcd. If a new write
+// needs a lease that has similar expiration time to a previous one, the old
+// lease will be reused to reduce the overhead of etcd, since lease operations
+// are expensive.
+type leaseManager struct {
+	// etcd client used to grant leases
+	client *clientv3.Client
+	// the number of buffered leases, i.e., size of the following mappings
+	size int
+	// the mutex used for the following maps
+	leaseMu sync.Mutex
+	// mapping from lease expiration time in seconds since 1970 to lease id
+	leaseCache map[int64]clientv3.LeaseID
+	// mapping form lease expiration time to last used time in seconds since 1970
+	leaseLastUsedTime map[int64]int64
```

Comment:
My personal feeling is that this is overkill.
Currently, TTL for all events (this is the only place where ttl is used) is constant. So introducing a complicated mechanism to support different ttls is a feature that:
- we will not take advantage of in any near term future
- it unnecessary complicates the code
- it makes it less performant (because choosing correct lease requires come computations)

I vote for getting rid of it and introducing an assumption that TTL is always constant (in fact I much preferred the previous solution with just a single cached lease).
@smarterclayton @liggitt @jpbetz @lavalamp @xiang90 for thoughts

Reply:
In the above discussion, @liggitt says there are other ttls. If others happen to interleave with events, the original simple design will not perform well.

Reply:
OK - apparently it's about openshift, not about Kubernetes itself.

Reply:
> apparently it's about openshift, not about Kubernetes itself.

or any consumer of `k8s.io/apiserver` that implements TTLFunc()

https://github.com/kubernetes/kubernetes/blob/bf422b7b04f9f8ab986de8ea016dbc5205c665a7/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L115-L121

Reply:
@wojtek-t

I discussed with @ccding offline, and suggested the initial solution with the assumption that TTL duration is a const (or mostly). 

With @liggitt's use case, I think the current solution is reasonable though more complicated.

Reply:
I thought the approach suggested in https://github.com/kubernetes/kubernetes/pull/64539#issuecomment-394057921 seemed reasonable for now (didn't require pooling/tracking, worst case devolved to current behavior)

Reply:
Okay, I will change it to the original simple solution.